proposals/p6395.md

```diff
@@ -0,0 +1,167 @@
+# Type completeness in extend
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/6395)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Abstract
+
+Require the facet type in `extend require impls` and `extend impl as`
+declarations to be complete, since `extend` changes the scopes which are looked
+in for name lookup.
+
+## Problem
+
+Proposal
+[#5168](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md)
+laid out rules for when a facet type needs to be identified or complete. When
+`require impls Z` is written, then `Z` must be identified. However it does not
+specify any rule for `extend require impls Z`. or `extend impl as Z`.
+
+If an interface `Y` uses `extend` for another interface `Z`, then a facet type
+for `Y` will also allow name lookup into `Z`. Similarly if a class `Y` uses
+`extend` for an interface `Z`, then an object of type `Y` will also allow name
+lookup into `Z`. In both cases, when the definition of `Y` is complete, we
+should also require `Z` to be complete, as its name scope becomes conceptually a
+part of `Y`.
+
+Additionally, after an `extend ... Z` declaration inside the definition of `Y`,
+the names of `Z` should be immediately available inside the `Y` definition,
+including inside the nested definition of `extend impl as Z { ... }`.
+
+## Background
+
+-   Proposal
+    [#5168](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md):
+    Forward impl declaration of an incomplete interface
+-   Proposal
+    [#2760](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2760.md):
+    Consistent class and interface syntax
+
+## Proposal
```

Comment:
Do we already have a rule for `extend base` and `extend adapt` declarations? If not, would it be worth adding them here too, so that we have complete coverage of all `extend` declarations?